src/pymor/operators/constructions.py

```diff
@@ -1430,3 +1430,90 @@ def apply_adjoint(self, V, mu=None):
+
+    Given an operator `A` acting on R^d, this class enables the functional::
+
+        op: R^d ----> R^1
+             u  |---> A(u, u)
+
+    In particular this is intended to provide the jacobian of such a bilinear
+    functional::
+
+        op.jacobian(U, mu).apply(V) = (A(u, v; mu) + A(v, u; mu))
+
+    Parameters
+    ----------
+    operator
+        The `Operator` to be wrapped as a bilinear functional.
+    """
+
+    linear = False
+    range = NumpyVectorSpace(1)
+
+    def __init__(self, operator, name=None):
+        assert operator.linear
+        assert operator.source == operator.range
+        self.__auto_init(locals())
+        self.source = operator.source
+
+    def apply(self, U, mu=None):
+        assert U in self.source
+        return self.range.from_numpy(self.operator.apply2(U, U, mu))
+
+    def jacobian(self, U, mu=None):
+        inner_vec = self.operator.apply_adjoint(U, mu) + self.operator.apply(U, mu)
+        return VectorFunctional(inner_vec, name=self.name + '_jacobian')
+
+    def d_mu(self, parameter, index=1):
+        # the parameter derivative only takes effect on the inner operator
+        return BilinearFunctional(self.operator.d_mu(parameter, index), name=self.name + '_d_mu')
+
+
+class BilinearProductFunctional(Operator):
+    """An `Operator` representing a bilinear functional by multiplying two linear functionals.
+
+    Given linear operators `A`, `B`: R^d ----> R^n, this class enables the functional::
+
+        op: R^d ----> R^1
+             u  |---> (A(u), B(u))
+
+    In particular this is intended to provide the jacobian of such a bilinear
+    functional::
+
+        op.jacobian(U, mu).apply(V) = (A(v; mu), B(u; mu)) + (A(u; mu), B(v; mu))
+
+    Parameters
+    ----------
+    operators
+        The `Operator`s to be wrapped as a bilinear functional.
+    """
+
+    linear = False
+    range = NumpyVectorSpace(1)
+
+    def __init__(self, operators, name=None):
+        assert len(operators) == 2
+        assert operators[0].source == operators[1].source
+        assert operators[0].range == operators[1].range
+        self.__auto_init(locals())
+        self.source = operators[0].source
+
+    def apply(self, U, mu=None):
+        assert U in self.source
+        AU = self.operators[0].apply(U, mu)
+        BU = self.operators[1].apply(U, mu)
+        return self.range.from_numpy(AU.pairwise_inner(BU).reshape((-1, 1)))
+
+    def jacobian(self, U, mu=None):
+        inner_vec = self.operators[1].apply_adjoint(self.operators[0].apply(U, mu), mu) \
+            + self.operators[0].apply_adjoint(self.operators[1].apply(U, mu), mu)
+        return VectorFunctional(inner_vec)
+
+    def d_mu(self, parameter, index=1):
+        # the parameter derivative only takes effect on the inner operators
+        ops = (self.operators[0].d_mu(parameter, index), self.operators[1].d_mu(parameter, index))
+        return BilinearProductFunctional(ops, name=self.name + '_d_mu')
```

Comment:
This seems wrong. In matrices one would have $o_\mu(u) = u^T B_\mu^T A_\mu u$, so the $\mu$-derivative should contain two summands. (If one were to use the construction suggested above, then this would be handled by `d_mu` of `ConcatenationOperator`.